presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java

```diff
@@ -389,6 +433,37 @@ else if (actualReturnType.isPrimitive()) {
 
             this.specializedTypeParameters = getDeclaredSpecializedTypeParameters(method);
 
+            allowedTypeParameters = ImmutableMap.<String, SubTypeSignature>builder()
```

Comment:
I would love to reuse the code if it exists somewhere to do the check. But somehow it is more complicated than I expect. For example, VarcharType does not have getTypeParameters(), so if the signature is "varchar(10, 3, 3)", we are unable to tell if it's legal or not.

Another case is that If the users specify something like @TypeParameter("varchar") or @TypeParameter("varchar(100)"), in both cases, the signature parsed back is going to be TYPE(LONG). But if it is @TypeParameter("char"), the signature would be TYPE only. It is quite bit annoying handle this case by case.